minecraft-common/src/main/java/com/sigmundgranaas/forgero/minecraft/common/toolhandler/VeinMiningStrategy.java

```diff
@@ -2,96 +2,111 @@
-            blockSet.add(rootPos);
-            queue.add(rootPos);
-            depth -= 1;
-        }
-
-        if (rootState.isIn(TagKey.of(Registry.BLOCK_KEY, new Identifier(handler.tag())))) {
-            calculateNextBlocks(blockSet, queue, depth, rootBlock, world, player);
-        }
-
-        for (BlockPos pos : blockSet) {
-            list.add(new Pair<>(world.getBlockState(pos), pos));
-        }
-
-        return list;
-    }
-
-    private void calculateNextBlocks(@NotNull HashSet<BlockPos> blockSet, @NotNull PriorityQueue<BlockPos> queue, int depth, @NotNull Block rootBlock, @NotNull BlockView world, @NotNull PlayerEntity player) {
-        if (depth < 1) {
-            return;
-        }
-
-        BlockPos pos = queue.peek();
-        if (pos == null) {
-            return;
-        }
-        queue.remove();
-        var directions = Arrays.asList(Direction.values());
-        //Reversing the direction to make the algorithm prioritize blocks in other direction that up and down
-        Collections.reverse(directions);
-        for (Direction direction : directions) {
-            BlockPos newBlock = pos.offset(direction, 1);
-
-            if (
-                    !blockSet.contains(newBlock)
-            ) {
-                if (world.getBlockState(newBlock).getBlock() == rootBlock) {
-                    depth -= 1;
-                    blockSet.add(newBlock);
-                    queue.add(newBlock);
-                }
-
-                for (Direction direction2 : directions) {
-                    BlockPos newBlock2 = newBlock.offset(direction2, 1);
-                    if (world.getBlockState(newBlock2).getBlock() == rootBlock && !blockSet.contains(newBlock2)) {
-                        depth -= 1;
-                        blockSet.add(newBlock2);
-                        queue.add(newBlock2);
-                    }
-                }
-                if (depth < 1) {
-                    return;
-                }
-            }
-        }
-        calculateNextBlocks(blockSet, queue, depth, rootBlock, world, player);
-    }
+	private final VeinBreaking handler;
 
+	private int currentDepth;
+
+	public VeinMiningStrategy(VeinBreaking handler) {
+		this.handler = handler;
+	}
+
+	public VeinMiningStrategy(PropertyData data) {
+		this.handler = new VeinBreaking((int) data.getValue(), data.getTags().get(0), data.getDescription());
+	}
+
+	@Override
+	public Set<BlockPos> getAvailableBlocks(BlockView world, BlockPos rootPos, PlayerEntity player) {
+		BlockState rootState = world.getBlockState(rootPos);
+		this.currentDepth = handler.depth();
+
+		if (BlockBreakingHandler.isBreakableBlock(world, rootPos, player) && rootState.isIn(TagKey.of(Registry.BLOCK_KEY, new Identifier(handler.tag())))) {
+			return getVeinMineableBlocksAtPosition(rootPos, world);
+		}
+
+		return Collections.emptySet();
+	}
+
+	private HashSet<BlockPos> getVeinMineableBlocksAtPosition(BlockPos rootBlockPos, BlockView world) {
```

Comment:
This is already a lot more clear and readable than the original one, good work!

Reply:
Thanks! I'll fix the rest of the things you mentioned soon:tm: